content/backend/graphql-scala/8-mutations.md

```diff
@@ -0,0 +1,315 @@
+
+```graphql
+
+mutation {
+  createUser(name: String!, authProvider: AuthProviderSignupData!): User
+}
+
+input AuthProviderSignupData {
+  email: AUTH_PROVIDER_EMAIL
+}
+
+input AUTH_PROVIDER_EMAIL {
+  email: String!
+  password: String!
+}
+```
+
+It isn't hard to imagine what this mutation does. Name suggests it's point of our interests - it creates an user, takes two paramaters of type `String` and `AuthProviderSignupData` and returns a `User`
+
+But wait... until now we've used `type` not `input`. So what is this? `input` is a type that can be used as parameter. You very often will see it along `mutation`s.
+
+Let's try to implement in order similar to creation of new query:
+
+* Define case classes for inputs
+* Define ObjectInputType's for those classes,
+* Define ObjectType responsible for all Mutations
+* Tell a Schema to use this object.
+
+
+### Create case classes
+
+<Instruction>
+
+Create classes we will need for inputs
+
+```scala
+
+case class AuthProviderEmail(email: String, password: String)
+
+case class AuthProviderSignupData(email: AuthProviderEmail)
+
+```
+
+</Instruction>
+
+### Define ObjectInputType's
+
+`ObjectInputType` is the same for input what `ObjectType` to `type` keyword.
+It tells Sangria how to understand a data. In fact you can define `ObjetType` and `ObjectInputType` for the same case class, or even more than one of each. Good example is an `User` entity which could consist many fields. But when you register new user and during signup action you need a different kind of data so you can create different ObjectInputTypes.
+
+<Instruction>
+
+In `GraphQLSchema.scala` add following definitions:
+
+```scala
+implicit val AuthProviderEmailInputType: InputObjectType[AuthProviderEmail] = deriveInputObjectType[AuthProviderEmail](
+   InputObjectTypeName("AUTH_PROVIDER_EMAIL")
+)
+
+implicit val AuthProviderSignupDataInputType: InputObjectType[AuthProviderSignupData] = deriveInputObjectType[AuthProviderSignupData]()
+
+```
+
+</Instruction>
+
+### Define Mutation Object
+
+It would be really similar to the process you already know.
+
+<Instruction>
+
+In the same file add a following code:
+
+```scala
+  val NameArg = Argument("name", StringType)
+  val AuthProviderArg = Argument("authProvider", AuthProviderSignupDataInputType)
+
+  val Mutation = ObjectType(
+    "Mutation",
+    fields[MyContext, Unit](
+      Field("createUser",
+        UserType,
+        arguments = NameArg :: AuthProviderArg :: Nil,
+        resolve = c => c.ctx.dao.createUser(c.arg(NameArg), c.arg(AuthProviderArg))
+      )
+    )
+  )
+```
+
+</Instruction>
+
+### Add Mutation to Schema
+
+<Instruction>
+
+Replace `schemaDefinition` with the code:
+
+```scala
+val SchemaDefinition = Schema(QueryType, Some(Mutation))
+```
+
+All mutations are optional so you have to wrap it in `Some`.
+
+If you will try to run a server now you will get errors about not implemented `FromInput`'s.
```

Comment:
+If you try to run a server, you will get errors about not implemented `FromInput`'s.